### PR TITLE
feat(auto-edit): Support unified diff and refactor diff format

### PR DESCRIPTION
**Pull request by umpox**
https://linear.app/sourcegraph/issue/CODY-4921/auto-edit-support-unified-diff-format-for-image-generations

https://github.com/user-attachments/assets/f9a0afdd-7e09-4cd7-85fe-53b99dc1d7fc

## Description

This PR:
- Adds support for a unified diff format with a different UI. This will be useful for other editors where we do not have support for decorations or will not have it for the MVP
- Refactors the image generator feature to use `DecorationInfo` instead of `AddedLinesDecorationInfo`. This means we can easily port this to the inline decorator in a follow up PR.
- Generally refactors much of the canvas/highlighting/diff logic. This was necessary to simplify this code as much as possible, as porting to `DecorationInfo` and supporting deletions here adds a lot of checks/edge cases.

## Test plan

- Manual testing with both `unified` and `additions` modes enabled
- Unit tests that generate images
- Further tests in this PR branched from here: https://github.com/sourcegraph/cody/pull/7138

### CODE REVIEW

File: vscode/src/autoedits/renderer/image-gen/decorated-diff/types.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * VisualDiff is an abstraction over DecoratedDiff that is suitable for rendering.
+ * It is an ordered list of relevant lines that we can simply iterate over and render.
+ *
+ * It supports `ModifiedLineInfoAdded` and `ModifiedLineInfoRemoved` to represent the two sides of a modified line.
+ * This is useful for rendering a unified diff
+ */
+export interface VisualDiff {
+    type: 'unified' | 'additions'
```
Comment (valerybugakov) [inline vscode/src/autoedits/renderer/image-gen/decorated-diff/types.ts:64]: Nit: we call it `mode` in the `SuggestionOptions` (`vscode/src/autoedits/renderer/image-gen/index.ts`). We could use a consistent name for it.

File: vscode/src/autoedits/renderer/image-gen/canvas/draw-decorations.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+export function drawDecorationsToCanvas(
+    diff: VisualDiff,
+    theme: SYNTAX_HIGHLIGHT_THEME,
+    mode: DiffMode,
+    userConfig: UserProvidedRenderConfig
+): EmulatedCanvas2D {
+    if (!canvasKit || !fontCache) {
+        // TODO: Log these errors, useful to see if we run into issues where we're not correctly
+        // initializing the canvas
+        throw new Error('Canvas not initialized')
```
Comment (valerybugakov) [inline vscode/src/autoedits/renderer/image-gen/canvas/draw-decorations.ts:114]: We should figure out why Sentry error reporting isn't working right now. It'd be helpful to check if this issue occurs on users' machines.

File: vscode/src/autoedits/renderer/image-gen/canvas/utils.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import type { VisualDiffLine } from '../decorated-diff/types'
+
+export function getRangesToHighlight(line: VisualDiffLine): [number, number][] {
+    switch (line.type) {
+        case 'removed':
```
Comment (valerybugakov) [inline vscode/src/autoedits/renderer/image-gen/canvas/utils.ts:5]: Nice! So easy to read this.

File: vscode/src/autoedits/renderer/image-gen/highlight/highlight-decorations.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    const syntaxHighlights: SyntaxHighlightRanges[] = []
+    let currentPosition = 0
+    for (const token of lineTokens) {
+        const tokenLength = token.content.length
+        const startPos = currentPosition
+        const endPos = currentPosition + tokenLength
+        syntaxHighlights.push({
+            range: [startPos, endPos],
+            color: token.color || DEFAULT_HIGHLIGHT_COLORS[theme],
+        })
```
Comment (valerybugakov) [inline vscode/src/autoedits/renderer/image-gen/highlight/highlight-decorations.ts:70]: The contrast on the light theme seems a bit low. Might be because I'm used to a higher-contrast color theme.

Screenshot 2025-02-20 at 16 14 15

It's a bit better for the dark theme:

Screenshot 2025-02-20 at 16 19 39

Comment (valerybugakov) [reply vscode/src/autoedits/renderer/image-gen/highlight/highlight-decorations.ts:70]: Yeah, it looks much better with the Monokai Dimmed dark theme:

Screenshot 2025-02-20 at 16 23 10

**Review by valerybugakov — APPROVED** [review]
Great work! Tested it locally and added a couple of minor suggestions inline.

**Comment by olafurpg** [comment]
@sourcegraph review

**Comment by umpox** [comment]
many thanks bot

File: vscode/src/autoedits/renderer/image-gen/highlight/highlight-decorations.ts (hunk 4 above)
Comment (umpox) [reply vscode/src/autoedits/renderer/image-gen/highlight/highlight-decorations.ts:70]: It's tricky because, the background image of the suggestion comes from your theme, but the syntax highlighting is hardcoded to a specific dark/light theme.

Some of the options we have:

1. Use a hardcoded background color for light/dark themes. I think we do this for chat right now. Solves the issue with a highlighting mismatch, but maybe becomes a visual annoyance in the users editor.

2. Find some way to get the users' theme colours. There's no built-in way that VS Code provides to do this, but maybe we could read the file system and the users settings to extract theme data.
